controllers/haproxyloadbalancer_controller.go

```diff
@@ -369,25 +368,118 @@ func (r haproxylbReconciler) BackEndpointsForCluster(ctx *context.HAProxyLoadBal
 			}
 		}
 
-		machineEndpoints := make([]corev1.EndpointAddress, 0)
-		for i, addr := range machine.Status.Addresses {
-			if addr.Type == clusterv1.MachineExternalIP {
-				// TODO(frapposelli): Remove this check once HAproxy fully supports IPv6 - issue #859
-				if utilnet.IsIPv6String(addr.Address) {
-					continue
-				}
-				endpoint := corev1.EndpointAddress{
-					NodeName: pointer.StringPtr(fmt.Sprintf("%s-%d", machine.Name, i)),
-					IP:       addr.Address,
-				}
-				machineEndpoints = append(machineEndpoints, endpoint)
-			}
-		}
+		machineEndpoints := getMachineEndpoints(*machine)
 		endpoints = append(endpoints, machineEndpoints...)
 	}
 	return endpoints, nil
 }
 
+// getMachineEndpoints gets the endpoints from machines
+func getMachineEndpoints(machine clusterv1.Machine) (machineEndpoints []corev1.EndpointAddress) {
+	for i, addr := range machine.Status.Addresses {
+		if addr.Type == clusterv1.MachineExternalIP {
+			// TODO(frapposelli): Remove this check once HAproxy fully supports IPv6 - issue #859
+			if utilnet.IsIPv6String(addr.Address) {
+				continue
+			}
+			endpoint := corev1.EndpointAddress{
+				NodeName: pointer.StringPtr(fmt.Sprintf("%s-%d", machine.Name, i)),
+				IP:       addr.Address,
+			}
+			machineEndpoints = append(machineEndpoints, endpoint)
+		}
+	}
+	return machineEndpoints
+}
+
+// extraServicesForCluster returns the extra services that the loadbalancer proxies to
+func (r haproxylbReconciler) extraServicesForCluster(ctx *context.HAProxyLoadBalancerContext) ([]haproxy.Frontend, error) {
+	frontEnds := make([]haproxy.Frontend, 0)
+
+	for _, extraService := range ctx.HAProxyLoadBalancer.Spec.ExtraServices {
+		// Get the CAPI Machine resources for the cluster.
+		machineList := &clusterv1.MachineList{}
+		if err := ctx.Client.List(
+			ctx, machineList,
+			ctrlclient.InNamespace(ctx.Cluster.Namespace),
+			ctrlclient.MatchingLabels(extraService.MachineSelector),
+		); err != nil {
+			return nil, errors.Wrap(err, "Failed to get machines for cluster")
```

Comment:
Additional info in error would be helpful. `"failed to get machines for cluster with labels %s"`

Reply:
Done